File: android/app/src/main/java/net/result/taulight/MethodHandlers.java
package net.result.taulight;

import android.annotation.TargetApi;
import android.os.Build;

import net.result.sandnode.exception.error.SandnodeErrorException;
import net.result.sandnode.link.Links;
import net.result.sandnode.link.SandnodeLinkRecord;
import net.result.sandnode.serverclient.SandnodeClient;
import net.result.sandnode.util.IOController;
import net.result.taulight.exception.ClientNotFoundException;

import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;

import java.util.*;
import java.util.Map.Entry;

import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.stream.Collectors;
import java.util.UUID;

import io.flutter.embedding.engine.FlutterEngine;
import io.flutter.plugin.common.BinaryMessenger;
import io.flutter.plugin.common.MethodCall;
import io.flutter.plugin.common.MethodChannel;

public class MethodHandlers {
    private static final Logger LOGGER = LogManager.getLogger(MethodHandlers.class);

    private interface MethodHandler {
        Object methodCall(MethodCall ignoredCall) throws Exception;
    }

    private final Map<String, MethodHandler> methodHandlerMap;
    private final Taulight taulight;

    private final ExecutorService executorService;

    private final Runner runner;

    public MethodHandlers(FlutterEngine flutterEngine) {
        executorService = Executors.newCachedThreadPool();

        BinaryMessenger binaryMessenger = flutterEngine.getDartExecutor().getBinaryMessenger();
        String CHANNEL = "net.result.taulight/messenger";

        MethodChannel methodChannel = new MethodChannel(binaryMessenger, CHANNEL);
        methodChannel.setMethodCallHandler(this::onMethodCallHandler);

        taulight = new Taulight(methodChannel);

        runner = new Runner(taulight);

        methodHandlerMap = new HashMap<>();
        methodHandlerMap.put("connect", this::connect);
        methodHandlerMap.put("login", this::login);
        methodHandlerMap.put("register", this::register);
        methodHandlerMap.put("disconnect", this::disconnect);
        methodHandlerMap.put("send", this::send);
        methodHandlerMap.put("group", this::groupAdd);
        methodHandlerMap.put("get-chats", this::getChats);
        methodHandlerMap.put("load-messages", this::loadMessages);
        methodHandlerMap.put("load-clients", this::loadClient);
        methodHandlerMap.put("load-chat", this::loadChat);
        methodHandlerMap.put("create-channel", this::createChannel);
        methodHandlerMap.put("members", this::members);
        methodHandlerMap.put("add-member", this::addMember);
        methodHandlerMap.put("token", this::token);
        methodHandlerMap.put("check-code", this::checkCode);
        methodHandlerMap.put("use-code", this::useCode);
        methodHandlerMap.put("dialog", this::dialog);
        methodHandlerMap.put("leave", this::leave);
        methodHandlerMap.put("channel-codes", this::channelCodes);
        methodHandlerMap.put("reply", this::reply);
    }

    @TargetApi(Build.VERSION_CODES.UPSIDE_DOWN_CAKE)
    private void onMethodCallHandler(MethodCall call, MethodChannel.Result result) {
        executorService.execute(() -> {
            MethodHandler handler = methodHandlerMap.get(call.method);
            if (handler != null) {
                Iterator<Entry<UUID, MemberClient>> iterator = taulight.clients.entrySet().iterator();
                while (iterator.hasNext()) {
                    Entry<UUID, MemberClient> entry = iterator.next();
                    if (!entry.getValue().client.io.isConnected()) {
                        LOGGER.debug("Removing client with uuid {}", entry.getKey());
                        iterator.remove();
                    }
                }

                try {
                    Object res = handler.methodCall(call);
                    result.success(Map.of("success", res));

                } catch (SandnodeErrorException e) {
                    Map<String, String> error = new HashMap<>();
                    error.put("name", e.getClass().getSimpleName());
                    error.put("message", e.getMessage());
                    result.success(Map.of("error", error));

                } catch (Exception e) {
                    LOGGER.error("Unhandled", e);
                    Map<String, String> error = new HashMap<>();
                    error.put("name", e.getClass().getSimpleName());
                    error.put("message", e.getMessage());
                    result.success(Map.of("error", error));
                }
            }
            else result.error("UNAVAILABLE", "Unknown method type", call.method);
        });
    }

    private Object members(MethodCall call) throws Exception {
        String uuid = call.argument("uuid");
        String chatIDStr = call.argument("chat-id");

        assert uuid != null;
        assert chatIDStr != null;

        SandnodeClient client = taulight.getClient(uuid).client;
        UUID chatID = UUID.fromString(chatIDStr);

        return runner.members(client, chatID);
    }

    private Map<String, String> connect(MethodCall call) throws Exception {
        String uuid = call.argument("uuid");
        String linkString = call.argument("link");

        assert uuid != null;
        assert linkString != null;

        SandnodeLinkRecord link = Links.parse(linkString);

        UUID clientID = UUID.fromString(uuid);

        return runner.connect(clientID, link);
    }

    private Map<String, String> login(MethodCall call) throws Exception {
        String uuid = call.argument("uuid");
        String nickname = call.argument("nickname");
        String password = call.argument("password");

        assert nickname != null;
        assert password != null;

        SandnodeClient client = taulight.getClient(uuid).client;

        return runner.login(client, nickname, password);
    }

    private Map<String, String> register(MethodCall call) throws Exception {
        String uuid = call.argument("uuid");
        String nickname = call.argument("nickname");
        String password = call.argument("password");

        assert nickname != null;
        assert password != null;

        SandnodeClient client = taulight.getClient(uuid).client;

        return runner.register(client, nickname, password);
    }

    private String disconnect(MethodCall call) throws ClientNotFoundException {
        String uuid = call.argument("uuid");
        return runner.disconnect(uuid);
    }

    @TargetApi(Build.VERSION_CODES.N)
    private String send(MethodCall call) throws Exception {
        String uuid = call.argument("uuid");
        String chatID = call.argument("chat-id");
        String content = call.argument("content");
        List<String> repliesString = call.argument("replies");

        assert uuid != null;
        assert chatID != null;
        assert content != null;
        assert repliesString != null;

        Set<UUID> replies = repliesString.stream()
                .map(UUID::fromString)
                .collect(Collectors.toSet());

        MemberClient mc = taulight.getClient(uuid);
        IOController io = mc.client.io;

        return runner.send(io, chatID, content, replies);
    }

    private String groupAdd(MethodCall call) throws Exception {
        String uuid = call.argument("uuid");
        String group = call.argument("group");

        assert uuid != null;
        assert group != null;

        SandnodeClient client = taulight.getClient(uuid).client;

        return runner.groupAdd(client, group);
    }

    /** @noinspection rawtypes*/
    @TargetApi(Build.VERSION_CODES.TIRAMISU)
    private List getChats(MethodCall call) throws Exception {
        String uuid = call.argument("uuid");
        assert uuid != null;

        SandnodeClient client = taulight.getClient(uuid).client;
        return runner.getChats(client);
    }

    private Map<String, Object> loadMessages(MethodCall call) throws Exception {
        String uuid = call.argument("uuid");
        String chatID_str = call.argument("chat-id");
        Integer index = call.argument("index");
        Integer size = call.argument("size");

        assert uuid != null;
        assert chatID_str != null;
        assert index != null;
        assert size != null;

        SandnodeClient client = taulight.getClient(uuid).client;
        UUID chatID = UUID.fromString(chatID_str);

        return runner.loadMessages(client, chatID, index, size);
    }

    @TargetApi(Build.VERSION_CODES.UPSIDE_DOWN_CAKE)
    private List<Map<String, String>> loadClient(MethodCall ignoredCall) {
        return runner.loadClient();
    }

    @TargetApi(Build.VERSION_CODES.N)
    private Object loadChat(MethodCall call) throws Exception {
        String uuid = call.argument("uuid");
        String chat = call.argument("chat-id");
        assert uuid != null;
        assert chat != null;

        UUID chatID = UUID.fromString(chat);
        SandnodeClient client = taulight.getClient(uuid).client;

        return runner.loadChat(client, chatID);
    }

    private Map<String, String> createChannel(MethodCall call) throws Exception {
        String uuid = call.argument("uuid");
        String title = call.argument("title");
        assert uuid != null;
        assert title != null;

        SandnodeClient client = taulight.getClient(uuid).client;

        return runner.createChannel(client, title);
    }

    @TargetApi(Build.VERSION_CODES.O)
    private Map<String, String> addMember(MethodCall call) throws Exception {
        String uuid = call.argument("uuid");
        String chatIDString = call.argument("chat-id");
        String otherNickname = call.argument("nickname");
        assert uuid != null;
        assert chatIDString != null;

        UUID chatID = UUID.fromString(chatIDString);

        SandnodeClient client = taulight.getClient(uuid).client;

        return runner.addMember(client, chatID, otherNickname);
    }

    private Map<String, String> token(MethodCall call) throws Exception {
        String uuid = call.argument("uuid");
        String token = call.argument("token");

        assert uuid != null;
        assert token != null;

        SandnodeClient client = taulight.getClient(uuid).client;

        return runner.token(client, token);
    }

    private Object checkCode(MethodCall call) throws Exception {
        String uuid = call.argument("uuid");
        String code = call.argument("code");

        assert uuid != null;
        assert code != null;

        SandnodeClient client = taulight.getClient(uuid).client;

        return runner.checkCode(client, code);
    }

    private String useCode(MethodCall call) throws Exception {
        String uuid = call.argument("uuid");
        String code = call.argument("code");

        assert uuid != null;
        assert code != null;

        SandnodeClient client = taulight.getClient(uuid).client;

        return runner.useCode(client, code);
    }

    private Map<String, String> dialog(MethodCall call) throws Exception {
        String uuid = call.argument("uuid");
        String nickname = call.argument("nickname");

        assert uuid != null;
        assert nickname != null;

        SandnodeClient client = taulight.getClient(uuid).client;

        return runner.dialog(client, nickname);
    }

    private String leave(MethodCall call) throws Exception {
        String uuid = call.argument("uuid");
        String chatIDStr = call.argument("chat-id");

        assert uuid != null;
        assert chatIDStr != null;

        UUID chatID = UUID.fromString(chatIDStr);
        SandnodeClient client = taulight.getClient(uuid).client;

        return runner.leave(client, chatID);
    }

    /** @noinspection rawtypes*/
    @TargetApi(Build.VERSION_CODES.N)
    private List channelCodes(MethodCall call) throws Exception {
        String uuid = call.argument("uuid");
        String chatIDStr = call.argument("chat-id");

        assert uuid != null;
        assert chatIDStr != null;

        UUID chatID = UUID.fromString(chatIDStr);
        SandnodeClient client = taulight.getClient(uuid).client;

        return runner.channelCodes(client, chatID);
    }

    @TargetApi(Build.VERSION_CODES.N)
    private String reply(MethodCall call) throws Exception {
        String uuid = call.argument("uuid");
        String chatIDStr = call.argument("chat-id");
        String content = call.argument("content");
        List<String> replyIDs = call.argument("replies");

        assert chatIDStr != null;
        assert content != null;

        MemberClient mc = taulight.getClient(uuid);


        UUID chatID = UUID.fromString(chatIDStr);
        Set<UUID> replies = new HashSet<>();

        if (replyIDs != null && !replyIDs.isEmpty()) {
            for (String replyID : replyIDs) {
                replies.add(UUID.fromString(replyID));
            }
        }

        return runner.reply(mc, chatID, content, replies);
    }
}

File: android/app/src/main/java/net/result/taulight/MethodHandlers.kt
package net.result.taulight

import android.os.Build
import android.util.Base64
import androidx.annotation.RequiresApi
import io.flutter.embedding.engine.FlutterEngine
import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import net.result.sandnode.chain.Chain
import net.result.sandnode.dto.FileDTO
import net.result.sandnode.exception.error.KeyStorageNotFoundException
import net.result.sandnode.link.Links
import net.result.sandnode.serverclient.SandnodeClient
import net.result.taulight.dto.ChatMessageInputDTO
import org.apache.logging.log4j.LogManager
import java.lang.reflect.InvocationTargetException
import java.util.*
import java.util.concurrent.ExecutorService
import java.util.concurrent.Executors

val uuidRegex = Regex("^[0-9a-fA-F]{8}-[0-9a-fA-F]{4}-[1-5][0-9a-fA-F]{3}-[89abAB][0-9a-fA-F]{3}-[0-9a-fA-F]{12}$")

val executorService: ExecutorService = Executors.newCachedThreadPool()

var methodHandlerMap: Map<String, (MethodCall) -> Any?> = emptyMap()

var taulight: Taulight? = null

object M {
    val LOGGER = LogManager.getLogger("MethodHandlers")!!
}

@RequiresApi(Build.VERSION_CODES.N)
class MethodHandlers(flutterEngine: FlutterEngine) {
    companion object {
        private const val CHANNEL = "net.result.taulight/messenger"
    }

    init {
        val binaryMessenger: BinaryMessenger = flutterEngine.dartExecutor.binaryMessenger
        val methodChannel = MethodChannel(binaryMessenger, CHANNEL)
        taulight = Taulight(methodChannel)

        methodChannel.setMethodCallHandler(this::onMethodCallHandler)
        methodHandlerMap = mapOf(
            "connect" to ::connect,
            "disconnect" to ::disconnect,
            "register" to ::register,
            "send" to ::send,
            "get-chats" to ::getChats,
            "load-messages" to ::loadMessages,
            "load-clients" to ::loadClients,
            "load-chat" to ::loadChat,
            "login-history" to ::loginHistory,
            "chain" to ::chain,
        )
    }

    fun onMethodCallHandler(call: MethodCall, result: MethodChannel.Result) {
        executorService.execute {
            val handler = methodHandlerMap[call.method]
            if (handler != null) {
                taulight!!.clients.entries.removeIf { (_, mc) -> !mc.client.io().isConnected }

                try {
                    val res: Any? = handler(call)
                    result.success(mapOf("success" to res))
                } catch (e: Exception) {
                    M.LOGGER.error("Unhandled", e)
                    result.success(
                        mapOf(
                            "error" to mapOf(
                                "name" to e.javaClass.simpleName,
                                "message" to e.message
                            )
                        )
                    )
                }
            } else result.error("UNAVAILABLE", "Unknown method type", call.method)
        }
    }
}

fun connect(call: MethodCall): Map<String, String> {
    val uuid = call.argument<String>("uuid")!!
    val clientID = UUID.fromString(uuid)

    val linkString = call.argument<String>("link")!!
    val link = Links.parse(linkString)

    val address = connect(taulight!!, clientID, link)

    return mapOf("address" to address.toString(52525))
}

fun disconnect(call: MethodCall): String {
    val uuid: String = call.argument<String>("uuid")!!
    disconnect(taulight!!, uuid)
    return "disconnected"
}

fun register(call: MethodCall): Map<String, String> {
    val uuid: String = call.argument<String>("uuid")!!
    val nickname: String = call.argument<String>("nickname")!!
    val password: String = call.argument<String>("password")!!
    val device: String = call.argument<String>("device")!!

    val client = taulight!!.getClient(uuid)

    val response = register(client.client, nickname, password, device)
    return mapOf(
        "token" to response.token,
        "key-id" to response.keyID.toString()
    )
}

fun send(call: MethodCall): Map<String, String> {
    val uuid: UUID = UUID.fromString(call.argument<String>("uuid")!!)
    val chatID: UUID = UUID.fromString(call.argument<String>("chat-id")!!)
    val content: String = call.argument<String>("content")!!
    val repliedToMessagesString: List<String> = call.argument<List<String>>("replied-to-messages")!!
    val fileIDsString: List<String> = call.argument<List<String>>("file-id")!!

    val mc: MemberClient = taulight!!.getClient(uuid)
    val repliedToMessages: Set<UUID> = repliedToMessagesString.map { UUID.fromString(it) }.toSet()
    val fileIDs: Set<UUID> = fileIDsString.map { UUID.fromString(it) }.toSet()

    val chat = taulight!!.getChat(chatID) ?: loadChat(mc.client, chatID)

    return send(mc.client, chat, content, repliedToMessages, fileIDs)
}

fun getChats(call: MethodCall): List<Map<String, Any>> {
    val uuid: String = call.argument<String>("uuid")!!
    val mc = taulight!!.getClient(uuid)
    return getChats(mc.client)
}

fun loadMessages(call: MethodCall): Map<String, Any> {
    val uuid: String = call.argument<String>("uuid")!!
    val chatIDStr: String = call.argument<String>("chat-id")!!
    val index: Int = call.argument<Int>("index")!!
    val size: Int = call.argument<Int>("size")!!

    val client: SandnodeClient = taulight!!.getClient(uuid).client
    val chatID: UUID = UUID.fromString(chatIDStr)

    val paginated = loadMessages(client, chatID, index, size)

    val messages = paginated.objects
    return mapOf(
        "count" to paginated.totalCount,
        "messages" to messages.map {
            val map: MutableMap<String, Any> = mutableMapOf(
                "message" to taulight!!.objectMapper.convertValue(it, Map::class.java)
            )

            try {
                val decrypted: String
                val input: ChatMessageInputDTO = it.message
                if (input.keyID != null) {
                    val agent = client.node().agent()
                    val keyStorage = agent.config.loadDEK(client.address, input.keyID)
                    val decoded = Base64.decode(input.content, Base64.NO_WRAP)
                    decrypted = keyStorage.encryption().decrypt(decoded, keyStorage)
                } else {
                    decrypted = input.content
                }

                map["decrypted"] = decrypted
            } catch (e: KeyStorageNotFoundException) {
                M.LOGGER.error("Send to flutter without decrypting - {}, {}", client, it, e)
            }

            map
        }
    )
}

@Suppress("unused")
fun loadClients(ignoredCall: MethodCall): List<Map<String, String>> {
    return taulight!!.clients.entries.map { (clientID, mc) ->
        val client = mc.client
        val map = mutableMapOf(
            "uuid" to clientID.toString(),
            "address" to client.address.toString(),
            "link" to mc.link.toString(),
        )

        if (client.nickname != null) map["nickname"] = client.nickname!!

        map
    }
}

fun loadChat(call: MethodCall): Map<String, Any> {
    val uuid: String = call.argument<String>("uuid")!!
    val chatString: String = call.argument<String>("chat-id")!!

    val chatID: UUID = UUID.fromString(chatString)
    val mc = taulight!!.getClient(uuid)
    val client = mc.client
    val chat = loadChat(client, chatID)

    val map: MutableMap<String, Any> = mutableMapOf(
        "chat" to taulight!!.objectMapper.convertValue(chat, Map::class.java)!!
    )

    chat.lastMessage?.let {
        try {
            decrypt(client, chat)
            map["decrypted-last-message"] = chat.decryptedMessage!!
        } catch (e: KeyStorageNotFoundException) {
            ChatRunner.LOGGER.error("Send to flutter without decrypting - {}, {}", client, chat, e)
        }
    }
    return map
}

fun loginHistory(call: MethodCall): List<Map<String, Any>> {
    val uuid: String = call.argument<String>("uuid")!!
    val mc = taulight!!.getClient(uuid)
    return loginHistory(mc.client)
}

fun chain(call: MethodCall): Any? {
    val uuid: String = call.argument<String>("uuid")!!
    val full: String = call.argument<String>("method")!!
    val params: List<Any> = call.argument<List<Any>>("params") ?: emptyList()

    val splitted = full.split(".")

    val className = splitted[0]
    val methodName = splitted[1]

    val client: SandnodeClient = taulight!!.getClient(uuid).client

    var clazz: Class<*>
    try {
        clazz = Class.forName("net.result.taulight.chain.sender.$className")
    } catch (_: ClassNotFoundException) {
        clazz = Class.forName("net.result.sandnode.chain.sender.$className")
    }
    val declaredConstructor = clazz.getDeclaredConstructor(SandnodeClient::class.java)
    val method = clazz.methods
        .firstOrNull { it.name == methodName && it.parameterTypes.size == params.size }
        ?: throw ClassNotFoundException()

    val chain: Chain? = declaredConstructor.newInstance(client) as Chain?

    client.io().chainManager.linkChain(chain)
    try {
        val processedParams = params.map {
            if (it is String && uuidRegex.matches(it)) {
                UUID.fromString(it)
            } else {
                it
            }
        }

        val result = method.invoke(chain, *processedParams.toTypedArray())
        return when (result) {
            Unit -> null
            is FileDTO -> mapOf(
                "id" to result.id().toString(),
                "contentType" to result.contentType(),
                "avatarBase64" to Base64.encodeToString(result.body(), Base64.NO_WRAP)
            )
            else -> taulight!!.objectMapper.convertValue(result, Any::class.java)
        }
    } catch (e: InvocationTargetException) {
        throw e.cause ?: e
    } finally {
        client.io().chainManager.removeChain(chain)
    }
}
